:bug: Teach Atom to read .git files

phord opened the pull request:
This fixes part of #8168.  The git-worktree model still is not supported, but this is because it is so new that libgit2 doesn't know about it yet, either.  But the .git-file feature which is heavily used by submodules works with this fix.

phord committed:
:white_check_mark: Add test for .git-files

phord committed:
:bug: Teach git-repository-provider to recognize .git-files

Git repositories may be contained in a .git directory or a .git file
in the workdir hierarchy, but Atom only recognizes the directory format.
Teach Atom to recognize the filesystem-agnostic Git symbolic link
used by default in many situations including, for example, submodules.

The .git file contains a relative or absolute path to the location of
the real git-dir, preceded by the 8-byte string "gitdir: ".

Here's a console log showing the normal creation of such a symbolic link.

    /tmp $ git init --separate-git-dir foo.git bar
    Initialized empty Git repository in /tmp/foo.git/
    /tmp $ ls
    /tmp $ bar  foo.git
    /tmp $ ls -la bar
    drwxr-xr-x 2 hordp hordp 4096 Sep 18 15:54 .
    drwxr-xr-x 4 hordp hordp 4096 Sep 18 15:54 ..
    -rw-r--r-- 1 hordp hordp   25 Sep 18 15:54 .git
    /tmp $ ls foo.git
    branches  config  description  HEAD  hooks  info  objects  refs
    /tmp $ cat bar/.git
    gitdir: /tmp/foo.git

Fixes #8876

maxbrunsfeld commented on spec/git-repository-provider-spec.coffee:
```diff
@@ -56,6 +56,21 @@ describe "GitRepositoryProvider", ->
+      it "returns a Promise that resolves to a GitRepository", ->
+        waitsForPromise ->
+          provider = new GitRepositoryProvider atom.project
+          gitDirPath = path.join(__dirname, 'fixtures/git/master.git')
```
For windows compatibility, I think this should be:

```coffee
path.join(__dirname, 'fixtures', 'git', 'master.git')
```

phord replied:
@maxbrunsfeld I expect you are correct.  But I copied that line from a previous test (line 27).

maxbrunsfeld replied:
I see. Sorry there are some bad examples in our test suite. Would you mind fixing this anyway, and adjusting the line that you copied it from in the same way?

phord replied:
I am happy to do so.

maxbrunsfeld commented:
@phord Other than my one small piece of feedback, this looks great. Thanks for diving in and fixing this!

phord committed:
Clean up *nix-specific paths in spec

There remain a few others in other specs, but someone who can test
them reliably should clean those up.
